contracts/fate.md

```diff
@@ -0,0 +1,819 @@
+# FATE
+The Fast æternity Transaction Engine
+
+## Design
+
+The high level machine (or the Fast æternity Transaction Engine) has
+aeternity transactions as its basic operations and it operates
+directly on the state tree of the aeternity chain. This is a new
+paradigm in blockchain virtual machine specifications which makes it
+possible to create type safe and efficient implementations of the
+machine.
+
+Every operation is typed and all values are typed (either by being
+stored in a typed memory or by a tag). Any type violation results in
+an exception and reverts all state changes. In version 1.0 there will
+be no catch instruction.
+
+In addition to normal machine instructions, such as ADD, the machine
+also has support for constructing most of the transactions available on
+the æternity chain from native low level chain transaction
+instructions.
+
+The instruction memory is divided into functions and basic
+blocks. Only basic blocks can be indexed and used as jump
+destinations.
+
+There are instructions to operate on the chain state tree in a safe and formalized way.
+
+FATE is "functional" in the sense that "updates" of data structures,
+such as tuples, lists or maps does not change the old values of the
+structure, instead a new version is created. Unless specific
+operations to write to the contract store are used.
+
+FATE does have the ability to write the value of an operation back to the
+same register or stack position as one of the arguments, in effect updating
+the memory. Still, any other references to the structure before the operation
+will have the same structure as before the operation.
+
+### Objectives
+
+#### Type safety
+
+FATE solves a fundamental problem programmers run into when coding for
+Ethereum: integer overflow, weak type checking and poor data
+flow. FATE checks all aritmetic operations to keep the right meanining
+of it. Also you can't implicitly cast types (eg integers to booleans).
+
+In EVM contracts are not typed. When calling a function, EVM will try
+to find which function you wanted to use by looking at the data that
+you send into the code. In FATE, you have actual functions and the
+functions have types - function call has to match the function type.
+
+FATE ultimately makes a safer coding platform for smart contracts.
+
+#### Rich data types
+
+FATE has more built in data types: like maps, lists.
+
+#### Faster transactions, smaller code size
+
+Having a higher level instructions makes the code deployed smaller and
+it reduces the blockchain size. Smaller code and smaller hashes also
+keeps a blockchain network from clogging up.
```

Comment:
Because of this, transactions also are cheaper. Is it worth mentioning it? I personally think that this is more important than the clogging statement.